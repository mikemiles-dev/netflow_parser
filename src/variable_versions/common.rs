use crate::protocol::ProtocolTypes;

use nom::bytes::complete::take;
use nom::error::{Error as NomError, ErrorKind};
use nom::number::complete::{be_u128, be_u32};
use nom::Err as NomErr;
use nom::IResult;
use nom_derive::*;
use serde::Serialize;

use std::net::{Ipv4Addr, Ipv6Addr};
use std::time::Duration;

/// Holds our datatypes and values post parsing
#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize)]
#[serde(untagged)]
pub enum DataNumber {
    U8(u8),
    U16(u16),
    U32(u32),
    U64(u64),
    U128(u128),
    I32(i32),
}

/// Convert into usize, mainly for serialization purposes
impl DataNumber {
    pub fn get_value(self) -> usize {
        let result: usize = self.into();
        result
    }

    /// Parse bytes into DataNumber Type
    pub fn parse(i: &[u8], field_length: u16, signed: bool) -> IResult<&[u8], DataNumber> {
        match field_length {
            1 if !signed => Ok(u8::parse(i)?).map(|(i, j)| (i, Self::U8(j))),
            2 if !signed => Ok(u16::parse(i)?).map(|(i, j)| (i, Self::U16(j))),
            4 if signed => Ok(i32::parse(i)?).map(|(i, j)| (i, Self::I32(j))),
            4 if !signed => Ok(u32::parse(i)?).map(|(i, j)| (i, Self::U32(j))),
            8 if !signed => Ok(u64::parse(i)?).map(|(i, j)| (i, Self::U64(j))),
            16 if !signed => Ok(u128::parse(i)?).map(|(i, j)| (i, Self::U128(j))),
            _ => Err(NomErr::Error(NomError::new(i, ErrorKind::Fail))),
        }
    }

    pub fn from_field_type(
        remaining: &[u8],
        field_type: FieldDataType,
        field_length: u16,
    ) -> IResult<&[u8], FieldValue> {
        let (remaining, field_value) = match field_type {
            FieldDataType::UnsignedDataNumber => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (i, FieldValue::DataNumber(data_number))
            }
            FieldDataType::SignedDataNumber => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, true)?;
                (i, FieldValue::DataNumber(data_number))
            }
            FieldDataType::String => {
                let (i, taken) = take(field_length)(remaining)?;
                (
                    i,
                    FieldValue::String(String::from_utf8_lossy(taken).to_string()),
                )
            }
            FieldDataType::Ip4Addr => {
                let (i, taken) = be_u32(remaining)?;
                let ip_addr = Ipv4Addr::from(taken);
                (i, FieldValue::Ip4Addr(ip_addr))
            }
            FieldDataType::Ip6Addr => {
                let (i, taken) = be_u128(remaining)?;
                let ip_addr = Ipv6Addr::from(taken);
                (i, FieldValue::Ip6Addr(ip_addr))
            }
            FieldDataType::DurationSeconds => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (
                    i,
                    FieldValue::Duration(Duration::from_secs(data_number.get_value() as u64)),
                )
            }
            FieldDataType::DurationMillis => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (
                    i,
                    FieldValue::Duration(Duration::from_millis(data_number.get_value() as u64)),
                )
            }
            FieldDataType::DurationMicros => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (
                    i,
                    FieldValue::Duration(Duration::from_micros(data_number.get_value() as u64)),
                )
            }
            FieldDataType::DurationNanos => {
                let (i, data_number) = DataNumber::parse(remaining, field_length, false)?;
                (
                    i,
                    FieldValue::Duration(Duration::from_nanos(data_number.get_value() as u64)),
                )
            }
            FieldDataType::ProtocolType => {
                let (i, protocol) = ProtocolTypes::parse(remaining)?;
                (i, FieldValue::ProtocolType(protocol))
            }
            FieldDataType::Float64 => {
                let (i, f) = f64::parse(remaining)?;
                (i, FieldValue::Float64(f))
            }
            FieldDataType::Vec => {
                let (i, taken) = take(field_length)(remaining)?;
                (i, FieldValue::Vec(taken.to_vec()))
            }
            FieldDataType::Unknown => {
                let (i, taken) = take(field_length)(remaining)?;
                (i, FieldValue::Vec(taken.to_vec()))
            }
        };
        Ok((remaining, field_value))
    }
}
/// Convert into usize, mainly for serialization purposes
impl From<DataNumber> for usize {
    fn from(val: DataNumber) -> Self {
        match val {
            DataNumber::U8(i) => i as usize,
            DataNumber::U32(i) => i as usize,
            DataNumber::I32(i) => i as usize,
            DataNumber::U16(i) => i as usize,
            DataNumber::U64(i) => i as usize,
            DataNumber::U128(i) => i as usize,
        }
    }
}

/// Holds the post parsed field with its relvant datatype
#[derive(Debug, PartialEq, PartialOrd, Clone, Serialize)]
pub enum FieldValue {
    String(String),
    DataNumber(DataNumber),
    Float64(f64),
    Duration(Duration),
    Ip4Addr(Ipv4Addr),
    Ip6Addr(Ipv6Addr),
    Vec(Vec<u8>),
    ProtocolType(ProtocolTypes),
    Unknown,
}

/// Helps the parser indefiy the data type to parse the field as
#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
pub enum FieldDataType {
    String,
    SignedDataNumber,
    UnsignedDataNumber,
    Float64,
    DurationSeconds,
    DurationMillis,
    DurationMicros,
    DurationNanos,
    Ip4Addr,
    Ip6Addr,
    Vec,
    ProtocolType,
    Unknown,
}
