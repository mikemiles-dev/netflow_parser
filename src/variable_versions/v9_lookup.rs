use super::common::*;

use nom_derive::*;
use serde::Serialize;

#[repr(u16)]
#[derive(Default, Debug, PartialEq, Eq, Clone, Copy, Serialize, Nom)]
pub enum ScopeFieldType {
    System = 1,
    Interface = 2,
    LineCard = 3,
    NetflowCache = 4,
    Template = 5,
    #[default]
    Unknown,
}

impl From<u16> for ScopeFieldType {
    fn from(item: u16) -> Self {
        match item {
            1 => ScopeFieldType::System,
            2 => ScopeFieldType::Interface,
            3 => ScopeFieldType::LineCard,
            4 => ScopeFieldType::NetflowCache,
            5 => ScopeFieldType::Template,
            _ => ScopeFieldType::Unknown,
        }
    }
}

#[repr(u16)]
#[derive(Debug, Hash, PartialEq, Eq, Clone, Ord, PartialOrd, Copy, Serialize, Nom)]
pub enum V9Field {
    InBytes = 1,
    InPkts = 2,
    Flows = 3,
    Protocol = 4,
    SrcTos = 5,
    TcpFlags = 6,
    L4SrcPort = 7,
    Ipv4SrcAddr = 8,
    SrcMask = 9,
    InputSnmp = 10,
    L4DstPort = 11,
    Ipv4DstAddr = 12,
    DstMask = 13,
    OutputSnmp = 14,
    Ipv4NextHop = 15,
    SrcAs = 16,
    DstAs = 17,
    BgpIpv4NextHop = 18,
    MulDstPkts = 19,
    MulDstBytes = 20,
    LastSwitched = 21,
    FirstSwitched = 22,
    OutBytes = 23,
    OutPkts = 24,
    MinPktLngth = 25,
    MaxPktLngth = 26,
    Ipv6SrcAddr = 27,
    Ipv6DstAddr = 28,
    Ipv6SrcMask = 29,
    Ipv6DstMask = 30,
    Ipv6FlowLabel = 31,
    IcmpType = 32,
    MulIgmpType = 33,
    SamplingInterval = 34,
    SamplingAlgorithm = 35,
    FlowActiveTimeout = 36,
    FlowInactiveTimeout = 37,
    EngineType = 38,
    EngineId = 39,
    TotalBytesExp = 40,
    TotalPktsExp = 41,
    TotalFlowsExp = 42,
    Ipv4SrcPrefix = 44,
    Ipv4DstPrefix = 45,
    MplsTopLabelType = 46,
    MplsTopLabelIpAddr = 47,
    FlowSamplerId = 48,
    FlowSamplerMode = 49,
    FlowSamplerRandomInterval = 50,
    MinTtl = 52,
    MaxTtl = 53,
    Ipv4Ident = 54,
    DstTos = 55,
    InSrcMac = 56,
    OutDstMac = 57,
    SrcVlan = 58,
    DstVlan = 59,
    IpProtocolVersion = 60,
    Direction = 61,
    Ipv6NextHop = 62,
    BpgIpv6NextHop = 63,
    Ipv6OptionHeaders = 64,
    MplsLabel1 = 70,
    MplsLabel2 = 71,
    MplsLabel3 = 72,
    MplsLabel4 = 73,
    MplsLabel5 = 74,
    MplsLabel6 = 75,
    MplsLabel7 = 76,
    MplsLabel8 = 77,
    MplsLabel9 = 78,
    MplsLabel10 = 79,
    InDstMac = 80,
    OutSrcMac = 81,
    IfName = 82,
    IfDesc = 83,
    SamplerName = 84,
    InPermanentBytes = 85,
    InPermanentPkts = 86,
    FragmentOffset = 88,
    ForwardingStatus = 89,
    MplsPalRd = 90,
    MplsPrefixLen = 91,
    SrcTrafficIndex = 92,
    DstTrafficIndex = 93,
    ApplicationDescription = 94,
    ApplicationTag = 95,
    ApplicationName = 96,
    PostipDiffServCodePoint = 98,
    Replicationfactor = 99,
    Deprecated = 100,
    Layer2packetSectionOffset = 102,
    Layer2packetSectionSize = 103,
    Layer2packetSectionData = 104,
    Vendor,
    Unknown,
}

impl From<V9Field> for FieldDataType {
    fn from(d: V9Field) -> FieldDataType {
        match d as u16 {
            1 => FieldDataType::UnsignedDataNumber,
            2 => FieldDataType::UnsignedDataNumber,
            3 => FieldDataType::UnsignedDataNumber,
            4 => FieldDataType::ProtocolType,
            5 => FieldDataType::UnsignedDataNumber,
            6 => FieldDataType::UnsignedDataNumber,
            7 => FieldDataType::UnsignedDataNumber,
            8 => FieldDataType::Ip4Addr,
            9 => FieldDataType::UnsignedDataNumber,
            10 => FieldDataType::UnsignedDataNumber,
            11 => FieldDataType::UnsignedDataNumber,
            12 => FieldDataType::Ip4Addr,
            13 => FieldDataType::UnsignedDataNumber,
            14 => FieldDataType::UnsignedDataNumber,
            15 => FieldDataType::Ip4Addr,
            16 => FieldDataType::UnsignedDataNumber,
            17 => FieldDataType::UnsignedDataNumber,
            18 => FieldDataType::Ip4Addr,
            19 => FieldDataType::UnsignedDataNumber,
            20 => FieldDataType::UnsignedDataNumber,
            21 => FieldDataType::DurationMillis,
            22 => FieldDataType::DurationMillis,
            23 => FieldDataType::UnsignedDataNumber,
            24 => FieldDataType::UnsignedDataNumber,
            25 => FieldDataType::UnsignedDataNumber,
            26 => FieldDataType::UnsignedDataNumber,
            27 => FieldDataType::Ip6Addr,
            28 => FieldDataType::Ip6Addr,
            29 => FieldDataType::UnsignedDataNumber,
            30 => FieldDataType::UnsignedDataNumber,
            31 => FieldDataType::UnsignedDataNumber,
            32 => FieldDataType::UnsignedDataNumber,
            33 => FieldDataType::UnsignedDataNumber,
            34 => FieldDataType::UnsignedDataNumber,
            35 => FieldDataType::UnsignedDataNumber,
            36 => FieldDataType::UnsignedDataNumber,
            37 => FieldDataType::UnsignedDataNumber,
            38 => FieldDataType::UnsignedDataNumber,
            39 => FieldDataType::UnsignedDataNumber,
            40 => FieldDataType::UnsignedDataNumber,
            41 => FieldDataType::UnsignedDataNumber,
            42 => FieldDataType::UnsignedDataNumber,
            44 => FieldDataType::Ip4Addr,
            45 => FieldDataType::Ip4Addr,
            46 => FieldDataType::UnsignedDataNumber,
            47 => FieldDataType::Ip4Addr,
            48 => FieldDataType::UnsignedDataNumber,
            49 => FieldDataType::UnsignedDataNumber,
            50 => FieldDataType::UnsignedDataNumber,
            52 => FieldDataType::UnsignedDataNumber,
            53 => FieldDataType::UnsignedDataNumber,
            54 => FieldDataType::UnsignedDataNumber,
            55 => FieldDataType::UnsignedDataNumber,
            56 => FieldDataType::UnsignedDataNumber,
            57 => FieldDataType::UnsignedDataNumber,
            58 => FieldDataType::UnsignedDataNumber,
            60 => FieldDataType::UnsignedDataNumber,
            61 => FieldDataType::UnsignedDataNumber,
            62 => FieldDataType::Ip6Addr,
            63 => FieldDataType::Ip6Addr,
            64 => FieldDataType::UnsignedDataNumber,
            70 => FieldDataType::UnsignedDataNumber,
            71 => FieldDataType::UnsignedDataNumber,
            72 => FieldDataType::UnsignedDataNumber,
            73 => FieldDataType::UnsignedDataNumber,
            74 => FieldDataType::UnsignedDataNumber,
            75 => FieldDataType::UnsignedDataNumber,
            76 => FieldDataType::UnsignedDataNumber,
            77 => FieldDataType::UnsignedDataNumber,
            78 => FieldDataType::UnsignedDataNumber,
            79 => FieldDataType::UnsignedDataNumber,
            80 => FieldDataType::UnsignedDataNumber,
            81 => FieldDataType::UnsignedDataNumber,
            82 => FieldDataType::UnsignedDataNumber,
            83 => FieldDataType::UnsignedDataNumber,
            84 => FieldDataType::UnsignedDataNumber,
            85 => FieldDataType::UnsignedDataNumber,
            86 => FieldDataType::UnsignedDataNumber,
            87 => FieldDataType::UnsignedDataNumber,
            88 => FieldDataType::UnsignedDataNumber,
            89 => FieldDataType::UnsignedDataNumber,
            90 => FieldDataType::Vec,
            91 => FieldDataType::UnsignedDataNumber,
            92 => FieldDataType::UnsignedDataNumber,
            93 => FieldDataType::UnsignedDataNumber,
            94 => FieldDataType::String,
            95 => FieldDataType::Vec,
            96 => FieldDataType::String,
            98 => FieldDataType::UnsignedDataNumber,
            99 => FieldDataType::UnsignedDataNumber,
            _ => FieldDataType::Unknown,
        }
    }
}

impl From<u16> for V9Field {
    fn from(item: u16) -> Self {
        match item {
            1 => V9Field::InBytes,
            2 => V9Field::InPkts,
            3 => V9Field::Flows,
            4 => V9Field::Protocol,
            5 => V9Field::SrcTos,
            6 => V9Field::TcpFlags,
            7 => V9Field::L4SrcPort,
            8 => V9Field::Ipv4SrcAddr,
            9 => V9Field::SrcMask,
            10 => V9Field::InputSnmp,
            11 => V9Field::L4DstPort,
            12 => V9Field::Ipv4DstAddr,
            13 => V9Field::DstMask,
            14 => V9Field::OutputSnmp,
            15 => V9Field::Ipv4NextHop,
            16 => V9Field::SrcAs,
            17 => V9Field::DstAs,
            18 => V9Field::BgpIpv4NextHop,
            19 => V9Field::MulDstPkts,
            20 => V9Field::MulDstBytes,
            21 => V9Field::LastSwitched,
            22 => V9Field::FirstSwitched,
            23 => V9Field::OutBytes,
            24 => V9Field::OutPkts,
            25 => V9Field::MinPktLngth,
            26 => V9Field::MaxPktLngth,
            27 => V9Field::Ipv6SrcAddr,
            28 => V9Field::Ipv4DstAddr,
            29 => V9Field::Ipv6SrcMask,
            30 => V9Field::Ipv6DstMask,
            31 => V9Field::Ipv6FlowLabel,
            32 => V9Field::IcmpType,
            33 => V9Field::MulIgmpType,
            34 => V9Field::SamplingInterval,
            35 => V9Field::SamplingAlgorithm,
            36 => V9Field::FlowActiveTimeout,
            37 => V9Field::FlowInactiveTimeout,
            38 => V9Field::EngineType,
            39 => V9Field::EngineId,
            40 => V9Field::TotalBytesExp,
            41 => V9Field::TotalPktsExp,
            42 => V9Field::TotalFlowsExp,
            43 => V9Field::Vendor,
            44 => V9Field::Ipv4SrcPrefix,
            45 => V9Field::Ipv4DstPrefix,
            46 => V9Field::MplsTopLabelType,
            47 => V9Field::MplsTopLabelIpAddr,
            48 => V9Field::FlowSamplerId,
            49 => V9Field::FlowSamplerMode,
            50 => V9Field::FlowSamplerRandomInterval,
            51 => V9Field::Vendor,
            52 => V9Field::MinTtl,
            53 => V9Field::MaxTtl,
            54 => V9Field::Ipv4Ident,
            55 => V9Field::DstTos,
            56 => V9Field::InSrcMac,
            57 => V9Field::OutDstMac,
            58 => V9Field::SrcVlan,
            59 => V9Field::DstVlan,
            60 => V9Field::IpProtocolVersion,
            61 => V9Field::Direction,
            62 => V9Field::Ipv6NextHop,
            63 => V9Field::BpgIpv6NextHop,
            64 => V9Field::Ipv6OptionHeaders,
            65 => V9Field::Vendor,
            66 => V9Field::Vendor,
            67 => V9Field::Vendor,
            68 => V9Field::Vendor,
            69 => V9Field::Vendor,
            70 => V9Field::MplsLabel1,
            71 => V9Field::MplsLabel2,
            72 => V9Field::MplsLabel3,
            73 => V9Field::MplsLabel4,
            74 => V9Field::MplsLabel5,
            75 => V9Field::MplsLabel6,
            76 => V9Field::MplsLabel7,
            77 => V9Field::MplsLabel8,
            78 => V9Field::MplsLabel9,
            79 => V9Field::MplsLabel10,
            80 => V9Field::InDstMac,
            81 => V9Field::OutSrcMac,
            82 => V9Field::IfName,
            83 => V9Field::IfDesc,
            84 => V9Field::SamplerName,
            85 => V9Field::InPermanentBytes,
            86 => V9Field::InPermanentPkts,
            87 => V9Field::Vendor,
            88 => V9Field::FragmentOffset,
            89 => V9Field::ForwardingStatus,
            90 => V9Field::MplsPalRd,
            91 => V9Field::MplsPrefixLen,
            92 => V9Field::SrcTrafficIndex,
            93 => V9Field::DstTrafficIndex,
            94 => V9Field::ApplicationDescription,
            95 => V9Field::ApplicationTag,
            96 => V9Field::ApplicationName,
            98 => V9Field::PostipDiffServCodePoint,
            99 => V9Field::Replicationfactor,
            100 => V9Field::Deprecated,
            102 => V9Field::Layer2packetSectionOffset,
            103 => V9Field::Layer2packetSectionSize,
            104 => V9Field::Layer2packetSectionData,
            _ => V9Field::Unknown,
        }
    }
}
