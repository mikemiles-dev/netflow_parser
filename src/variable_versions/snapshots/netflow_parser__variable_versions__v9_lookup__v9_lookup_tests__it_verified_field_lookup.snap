---
source: src/variable_versions/v9_lookup.rs
expression: fields
---
- InBytes
- InPkts
- Flows
- Protocol
- SrcTos
- TcpFlags
- L4SrcPort
- Ipv4SrcAddr
- SrcMask
- InputSnmp
- L4DstPort
- Ipv4DstAddr
- DstMask
- OutputSnmp
- Ipv4NextHop
- SrcAs
- DstAs
- BgpIpv4NextHop
- MulDstPkts
- MulDstBytes
- LastSwitched
- FirstSwitched
- OutBytes
- OutPkts
- MinPktLngth
- MaxPktLngth
- Ipv6SrcAddr
- Ipv6DstAddr
- Ipv6SrcMask
- Ipv6DstMask
- Ipv6FlowLabel
- IcmpType
- MulIgmpType
- SamplingInterval
- SamplingAlgorithm
- FlowActiveTimeout
- FlowInactiveTimeout
- EngineType
- EngineId
- TotalBytesExp
- TotalPktsExp
- TotalFlowsExp
- Vendor
- Ipv4SrcPrefix
- Ipv4DstPrefix
- MplsTopLabelType
- MplsTopLabelIpAddr
- FlowSamplerId
- FlowSamplerMode
- FlowSamplerRandomInterval
- Vendor
- MinTtl
- MaxTtl
- Ipv4Ident
- DstTos
- InSrcMac
- OutDstMac
- SrcVlan
- DstVlan
- IpProtocolVersion
- Direction
- Ipv6NextHop
- BpgIpv6NextHop
- Ipv6OptionHeaders
- Vendor
- Vendor
- Vendor
- Vendor
- Vendor
- MplsLabel1
- MplsLabel2
- MplsLabel3
- MplsLabel4
- MplsLabel5
- MplsLabel6
- MplsLabel7
- MplsLabel8
- MplsLabel9
- MplsLabel10
- InDstMac
- OutSrcMac
- IfName
- IfDesc
- SamplerName
- InPermanentBytes
- InPermanentPkts
- Vendor
- FragmentOffset
- ForwardingStatus
- MplsPalRd
- MplsPrefixLen
- SrcTrafficIndex
- DstTrafficIndex
- ApplicationDescription
- ApplicationTag
- ApplicationName
- Unknown
- PostipDiffServCodePoint
- Replicationfactor
- Deprecated
- Unknown
- Layer2packetSectionOffset
- Layer2packetSectionSize
- Layer2packetSectionData
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- IcmpTypeValue
- IcmpCodeValue
- IcmpIpv6TypeValue
- ImpIpv6CodeValue
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- PostNATSourceIPv4Address
- PostNATDestinationIPv4Address
- PostNATTSourceTransportPort
- PostNATTDestinationTransportPort
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- Unknown
- PostNATSourceIpv6Address
- PostNATDestinationIpv6Address
